STT GCP implementation

cvhooser opened the pull request:
Basic GCP implementation

**TODO**
1. Implement Test Cases for GCP
2. Get the v1 and v1beta speech clients working together simultaneously to take the more accuracte.
- I was trying to send each chunk to both the v1 and v1beta clients since I noticed they have different outputs (we are probably better off with the v1 version as of now). But I want to be able to submit both and then take the result that has the higher confidence rating. It would be sending double requests to the api though, so that is worth keeping in mind.

**Notes:** I worked on different chunking techniques with bufferOnSilence (don't think this was working right), bufferLimit, and bufferCount. I ended up leaving a `bufferCount(3)` to send slightly larger audio snippets. I did see some improvements in translation though maybe not enough to keep it long term. I still need to do more testing. No other chunking proved to be a measurable over the default chunking that is used.

cvhooser committed:
feat: GCP Raw Implementation

- Documentation Fixes
- Formatting consistent to style guide
- Added a clean script to the package.json
- Added Sample Audio Files
- GCP chunking without streaming or socket connection
*Note*: pipe(map, concatMap, map) has the effect of an indexed MergeMap or an async ConcatMap

cvhooser committed:
Merge branch 'dev' into s2t

cvhooser committed:
Merge branch 's2t' into st2-gcp

# Conflicts:
#	packages/rxtranscribe/demo/runDemo.js

cvhooser committed:
Slack / Cercle build integration

cvhooser committed:
Build Error Fixes

- Dependency Fixes
- Linting
- Moving audio samples to ‘samples’ folder
- Fixing toAws.test.js

cvhooser committed:
Adding a bufferCount for slightly higher accuracy

- Lots of playing with vad and bufferBetweenSilences
- No success with either

cvhooser commented:
![](https://github.trello.services/images/mini-trello-icon.png) [txtranscribe v1 (GCP)](https://trello.com/c/os1Fwqhl/48-txtranscribe-v1-gcp)

bfla commented on packages/rxtranscribe/demo/runDemo.js:
```diff
@@ -6,28 +6,27 @@ const {tap} = require('rxjs/operators');
   switch (strategy) {
     case 'deepspeech':
-      return toDeepSpeech({modelDir, codec: 'pcm'});
+      return toDeepSpeech({modelDir});
```
❓ I should double-check this...  Not sure if it's the same as what I have or not...

cvhooser replied:
Might have been some incorrect merging that I didn't notice. Good spot.

bfla commented on packages/rxtranscribe/demo/runDemo.js:
```diff
@@ -76,6 +85,8 @@ prompt.get(schema, (err, params) => {
+
   const transcription$ = runDemo(params);
   transcription$.subscribe(
     out => console.log(out),
```
👍 Nice! Looks good

bfla commented on .circleci/config.yml:
```diff
@@ -32,6 +33,9 @@ jobs:
           command: yarn build
+      - slack/status:
+          fail_only: true
+          failure_message: ':red_circle: A $CIRCLE_JOB has failed! You should probably do something about that...'
```
🤯  Mind blown. Cool trick

cvhooser replied:
We can do so much more with these orbs too! This was the simplest implementation that I did just for prototyping after I pushed a failed build.

bfla commented on package.json:
```diff
@@ -13,7 +13,8 @@
     "release": "lerna publish",
-    "release:beta": "lerna publish --canary --preid beta"
+    "release:beta": "lerna publish --canary --preid beta",
+    "clean": "lerna clean && rm -r node_modules"
```
👏 Good idea

bfla commented on packages/rxtranscribe/src/operators/bufferBetweenSilences.js:
```diff
@@ -10,29 +10,30 @@
+//       // tap(b => console.log('BUFFER_OUT', b))
 //     );
 //     return buffer$;
 //   };
```
🤔 Cool.  I've never seen `mergeAll` before. 

❓ Were you able to get `node-vad` working on your local?  I was having some issues getting it to install properly.  Hopefully it's not one of those crappy packages where you have to do a global install of some other package to make the npm library work...

bfla replied:
Uh wow. `mergeAll` is great.  I can't believe I missed that one in the RxJS docs.  The fact that you can control concurrency with it is great

cvhooser replied:
I had no issues getting it installed and running, but I couldn't get it working properly. There were no improvements in my workflow when using it despite numerous efforts. The `VAD` tests passed when I had it uncommented and ran them. The only difference that `bufferOnSilence` had, was to ignore the first chunk of data. Resulting in losing the first sentence from my audio file.

cvhooser replied:
I've been debugging `bufferOnSilence` for a couple hours now trying to get it to work with `GCP STT` and I think I figured out the problem. `VAD` works best on audio buffers between 26-144ms, yet we are feeding it chunks that are 25-150 times that size! We are effectively having VAD perform a useless operation. This means that the `bufferOnSilence` needs to take whatever chunk it was fed, chop off the last couple seconds and feed it into VAD until we have `x` milliseconds of consecutive silence and then return that partitioned chunk while simultaneously re-buffering the remaining amount so it can be prepended to the next incoming chunk. In the case of no more incoming chunks the last bit of data that was supposed to be prepended to the next incoming chunk is emitted.

bfla replied:
Oh... That explains it!

bfla commented on packages/rxtranscribe/src/operators/toGCP.js:
```diff
@@ -0,0 +1,65 @@
+        !googleCreds
+        ? throwError(new Error('Google Application Credentials must be set'))
+        : of(fileChunk)
+      )),
```
💡 I was doing this a lot too in my code but I realized later that something like this saves a few lines of code and accomplishes the same result:
```js
return fileChunk => {
  if (!googleCreds) return throwError(new Error('Google Application Credentials must be set'));
  // ... otherwise, run the normal pipeline
}
```

cvhooser replied:
I'm always down for more concise code. Good tip.

bfla commented on packages/rxtranscribe/src/operators/toGCP.js:
```diff
@@ -0,0 +1,65 @@
+      )),
+      bufferCount(3),
+      // bufferBetweenSilences({timeInterval: 1000, vadOptions: { sampleRate }}),
+      mergeAll(),
```
Curious -- what does `mergeAll` do here?  I'm not sure I understand what `mergeAll` does.  I thought it was for flattening an observable of observables into a single stream but the output of the previous step seems to be an array (each of 3 file chunks)?

cvhooser replied:
`bufferCount(3)` creates an array of 3 chunks (each arrays  of bytes) and then `mergeAll()` takes that array of arrays and flattens it creating a chunk that is three times the length of the original to be sent through the engine. `MergeMap` is what flattens an observable of observables.

cvhooser replied:
Further down in the commented out code you can see how I'm using `MergeMap` with `ForkJoin` to resolve an observable of observables. I was sad I couldn't get the two different clients working simultaneously there...

bfla commented on packages/rxtranscribe/src/operators/toGCP.js:
```diff
@@ -0,0 +1,65 @@
+      //   );
+      // }),
+      map(fileChunk => clientv1.recognize({config, audio: { content: fileChunk.toString('base64') }})),
+      concatMap(r => r),
```
😄  I'm having fun reading your code.  I'm learning some neat tricks.  So `concatMap` enforces order to ensure that each piece of text is processed before the next is sent?  That's a cool use case for it.  It didn't occur to me that concatMap could produce that result and that's pretty handy!

cvhooser replied:
Precisely. It allows me to initiate all requests at the same time while maintaining order. I have the speed of parallel processing with the ensured order of sequential processing.

bfla replied:
🤯

cvhooser commented on packages/rxws/src/operators/conduit.js:
```diff
@@ -1,4 +1,8 @@
-import {merge,of,throwError} from 'rxjs';
```
Whoops this was unecessary

cvhooser committed:
Bug fix and word confidence enabled

Switched to the v1beta to enable word level confidence scores

Fixed a bug with buffers not being concated which was causing accuracy to be static regardless of the stream size send to to the gcp stt client.

Removed excess code

cvhooser committed:
Added more metadata and switched to a more precise model

bfla committed:
Merge branch 'dev' into st2-gcp

cvhooser committed:
Merge branch 'dev' into st2-gcp

# Conflicts:
#	packages/rxtranscribe/demo/runDemo.js
#	packages/rxtranscribe/package-lock.json
